File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int sz = lists.size();
        ListNode dummy;
        ListNode* temp = &dummy;

        vector<pair<int, int>> heap;
        for (int i = 0; i < sz; ++i) {
            if (lists[i]) {
                heap.push_back(make_pair(lists[i]->val, i));
            }
        }
        make_heap(heap.begin(), heap.end(), greater<pair<int, int>>());
        
        while (heap.size() > 0) {
            const pair<int, int> kv = heap[0];
            pop_heap(heap.begin(), heap.end(), greater<pair<int, int>>());
            heap.pop_back();

            int node_val = kv.first;
            int list_index = kv.second;
            temp->next = lists[list_index];
            temp = temp->next;
            lists[list_index] = lists[list_index]->next;

            if (lists[list_index]) {
                heap.push_back(make_pair(lists[list_index]->val, list_index));
                push_heap(heap.begin(), heap.end(), greater<pair<int, int>>());
            }
        }
        temp->next = nullptr;

        return dummy.next;
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        cpy = head
        size = 0
        while cpy:
            size += 1
            cpy = cpy.next
        
        dummy = ListNode()
        dummy.next = head
        cpy = dummy
        for _ in range(size - n):
            cpy = cpy.next
        cpy.next = cpy.next.next
        return dummy.next